### ISSUE TITLE
[BUG] Bad date parsing

### ISSUE THREAD

**Issue by eddi13** [barbushin/php-imap#374]
**Environment (please complete the following information):**
 - PHP IMAP version: 3.0.28
 - PHP Version: 7.2.21
 - Type of execution: CLI or Web Server

**Describe the bug**
Dates without specifying the day of the week are not parsed

The headers of the parsed email:
```
Received: (qmail 28902 invoked for bounce); 21 Aug 2019 05:51:51 -0000
Date: 21 Aug 2019 05:51:51 -0000
From: MAILER-DAEMON@yahoo.com
```

https://tools.ietf.org/html/rfc2822#section-3.3
As described there day of the week is optional and may be absent. And the header of Yahoo is valid.

My suggestion for Mailbox::parseDateTime(). Lines 1189-1190:
```
//date can be:
//Wed, 21 Aug 2019 07:49:04 +0200 (CEST)
//Wed, 21 Aug 2019 07:49:04 +0200
//21 Aug 2019 07:49:04 +0200
//we crop " (CEST)", and, because day of the week is
//optional and may be absent, we crop it also
//from DateTime::RFC2822 ('D, d M Y H:i:s O') remains only 'd M Y H:i:s O'
$dateRegex = '/^[a-z]+,\s*|\s*\(.*\)$/i';
$dateFormatted = DateTime::createFromFormat('d M Y H:i:s O', preg_replace($dateRegex, '', trim($dateHeader)));
```
best regards
Eduard

**Comment by eddi13** [comment]
Sorry, I'm forgot.
The date is set in IncomingMailHeader::date parameter without TimeZone. That's why you can not say exactly when the mail was sent.

My suggestion:
Because the DateTime::RFC2822 standard is preferred for the data, the IncomingMailHeader::date can be formatted as well.
e.g. Wed, 21 Aug 2019 07:49:04 0200

best regards
Eduard

**Comment by eddi13** [comment]
I have now a date: Wed, 21 Aug 2019 07:49:04 KST +0900
we can not get by with the cut

Another suggestion:
If we do not cut, but choose us what we need
```
$hDate = date from header
$headerDate = '';
$dateFormat = 'd M Y H:i:s O';
$dateRegex = '/(?P<d>[0-9]{1,2})\s*(?P<M>[A-Z][a-z]{2})\s*(?P<Y>[0-9]{4})\s*(?P<H>[0-9]{2}):(?P<i>[0-9]{2})(?::(?P<s>[0-9]{2}))?.*?(?P<O>[+-]?[0-9]{4})/';
if($hDate && preg_match($dateRegex, $hDate, $maches)) {
	$search = [
		'd',
		'M',
		'Y',
		'H',
		'i',
		's',
		'O',
	];
	$replace = [
		$maches['d'],
		$maches['M'],
		$maches['Y'],
		$maches['H'],
		$maches['i'],
		$maches['s'] ?: '00',//seconds is optional in rfc2822
		$maches['O'],
	];
	$headerDate = str_replace($search, $replace, $dateFormat); //we get here the date for DateTime :: createFromFormat ()
}
```
I hope that could help

best regards
Eduard

**Comment by Sebi94nbg** [comment]
You're right. Those dates, which have no day-of-the-week in it, are parsed incorrectly. I've added some test cases for this issue, but I couldn't figure out a solution yet.

You suggested code did unfortunately not solve the issue. Feel free to test it by your own.

**Comment by eddi13** [comment]
I do not know what you mean by "not solve the issue".
Here's the regex test: https://regex101.com/r/XRd2qu/1/
Or so: https://regex101.com/r/XRd2qu/2/ but here is 100 steps more.
Here is my code. I have received ~15000 messages and have no parsing exceptions:
```
class SNIncomingMail extends IncomingMail {
	...
	public function getMessageDate() {
		if($this->messageDate === false) {
			$head = $this->headers;
			$hDate = '';
			if(isset($head->date) && !empty($head->date)) {
				$hDate = $head->date;
			}
			elseif(isset($head->Date) && !empty($head->Date)) {
				$hDate = $head->Date;
			}
			$headerDate = '';
			$dateFormat = 'd M Y H:i:s O';
			$regex = '/(?P<d>[0-9]{1,2})\s*(?P<M>[A-Z][a-z]{2})\s*(?P<Y>[0-9]{4})\s*(?P<H>[0-9]{2}):(?P<i>[0-9]{2})(?::(?P<s>[0-9]{2}))?.*?(?P<O>[+-]?[0-9]{4})/';
			if($hDate && preg_match($regex, $hDate, $maches)) {
				$search = [
					'd',
					'M',
					'Y',
					'H',
					'i',
					's',
					'O',
				];
				$replace = [
					$maches['d'],
					$maches['M'],
					$maches['Y'],
					$maches['H'],
					$maches['i'],
					$maches['s'] ?: '00', //seconds is optional
					$maches['O'],
				];
				$headerDate = str_replace($search, $replace, $dateFormat);
			}
			elseif($hDate) {
				//log parsing errors
				BugHandler::logException(new ExceptionSN("SNIncomingMail::getMessageDate() cannot parse header date '$hDate'"));
			}
			//this is not relevant. SNDate extends DateTime, it returns NULL if the date can not be created
			$this->messageDate = $headerDate ? SNDate::createFromFormat($dateFormat, $headerDate) : null;
		}
		return $this->messageDate;
	}
 }
```

**Comment by Sebi94nbg** [comment]
I've updated the code here: [L1176-L1199](https://github.com/barbushin/php-imap/blob/master/src/PhpImap/Mailbox.php#L1176-L1199) (not committed)

And then I've tested it: `phpunit --testdox`

The test cases failed for more then with the current code.

This is, what I mean with "does not solve the issue".

You can try to update the linked function `parseDateTime($dateHeader)` by your own and test it. If it's working for you, feel free to submit a Pull Request (PR).

**Comment by eddi13** [comment]
Thanks for the answer, but with PR it gets difficult because for the time being I disagree with logic "if nothing then now" at date parsing, in further, IncomingMailHeader::date is without Tinezone, that's not enough for my purposes, then, I need the date in my SNDate format.
I do not want to mix that deeply with your logic, so I've written my own method, and I do not use IncomingMailHeader::date at all.
To do that, let the class get a couple of settings, like $dateTimeClass and $dateFormat.
The meaning of my Issue was to show you that date parsing is not optimal.
How can I help right now: I am processing ~1000 messages per day. If my method fails, I'll contact you.

**Comment by Sebi94nbg** [comment]
This library tries to format the given header datetime to a RFC3339 compliant format. If this fails, it will return the original value.

Released in version [3.0.31](https://github.com/barbushin/php-imap/releases/tag/3.0.31).
